go/vt/vttablet/tabletserver/schema/historian.go

```diff
@@ -0,0 +1,286 @@
+	Reload(ctx context.Context) error
+	GetConnection(ctx context.Context) (*connpool.DBConn, error)
+	GetSchema() map[string]*Table
+}
+
+// TrackedSchema has the snapshot of the table at a given pos (reached by ddl)
+type TrackedSchema struct {
+	schema map[string]*binlogdatapb.MinimalTable
+	pos    mysql.Position
+	ddl    string
+}
+
+var _ Historian = (*HistorianSvc)(nil)
+
+// HistorianSvc implements the Historian interface by calling schema.Engine for the underlying schema
+// and supplying a schema for a specific version by loading the cached values from the schema_version table
+// The schema version table is populated by the Tracker
+type HistorianSvc struct {
+	he                  HistoryEngine
+	lastID              int64
+	schemas             []*TrackedSchema
+	mu                  sync.Mutex
+	trackSchemaVersions bool
+	lastSchema          map[string]*binlogdatapb.MinimalTable
+	isOpen              bool
+}
+
+// Reload gets the latest schema from the database
+func (h *HistorianSvc) Reload(ctx context.Context) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	return h.reload(ctx)
+}
+
+// NewHistorian creates a new historian. It expects a schema.Engine instance
+func NewHistorian(he HistoryEngine) *HistorianSvc {
+	sh := HistorianSvc{he: he, lastID: 0, trackSchemaVersions: true}
+	return &sh
+}
+
+// SetTrackSchemaVersions can be used to turn on/off the use of the schema_version history in the historian
+// Only used for testing
+func (h *HistorianSvc) SetTrackSchemaVersions(val bool) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.trackSchemaVersions = val
+}
+
+// readRow converts a row from the schema_version table to a TrackedSchema
+func (h *HistorianSvc) readRow(row []sqltypes.Value) (*TrackedSchema, int64, error) {
+	id, _ := evalengine.ToInt64(row[0])
+	pos, err := mysql.DecodePosition(string(row[1].ToBytes()))
+	if err != nil {
+		return nil, 0, err
+	}
+	ddl := string(row[2].ToBytes())
+	timeUpdated, _ := evalengine.ToInt64(row[3])
+	sch := &binlogdatapb.MinimalSchema{}
+	if err := proto.Unmarshal(row[4].ToBytes(), sch); err != nil {
+		return nil, 0, err
+	}
+	log.Infof("Read tracked schema from db: id %d, pos %v, ddl %s, schema len %d, time_updated %d \n",
+		id, mysql.EncodePosition(pos), ddl, len(sch.Tables), timeUpdated)
+
+	tables := map[string]*binlogdatapb.MinimalTable{}
+	for _, t := range sch.Tables {
+		tables[t.Name] = t
+	}
+	trackedSchema := &TrackedSchema{
+		schema: tables,
+		pos:    pos,
+		ddl:    ddl,
+	}
+	return trackedSchema, id, nil
+}
+
+// loadFromDB loads all rows from the schema_version table that the historian does not have as yet
+func (h *HistorianSvc) loadFromDB(ctx context.Context) error {
+	conn, err := h.he.GetConnection(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Recycle()
+	log.Infof("In loadFromDb for lastID %d", h.lastID)
+	tableData, err := conn.Exec(ctx, fmt.Sprintf(getSchemaVersions, h.lastID), 10000, true)
+	if err != nil {
+		log.Infof("Error reading schema_tracking table %v", err)
```

Comment:
Let's add a further qualification here: "will operate with the latest available schema". This should remain Infof logging.